shell/packages/sandstorm-ui-powerbox/powerbox-client.js

```diff
@@ -0,0 +1,445 @@
+SandstormPowerboxRequest = class SandstormPowerboxRequest {
+  constructor(db, requestInfo) {
+    check(requestInfo, Match.ObjectIncluding({
+      source: Match.Any,
+      origin: Match.Any,
+      rpcId: Match.Any,
+      // For reasons I don't understand, Match.Optional does not work here.
+      query: Match.OneOf(undefined, [Object]),
+      saveLabel: Match.Optional(Match.Any),
+      grainId: String,
+      identityId: String,
+      onCompleted: Function,
+    }));
+
+    this._db = db;
+    this._requestInfo = requestInfo;
+    this._completed = false;
+    this._error = new ReactiveVar(undefined);
+
+    // State to track UI status.
+    this._filter = new ReactiveVar("");
+    this._selectedProvider = new ReactiveVar(undefined);
+  }
+
+  finalize() {
+    if (!this._completed) {
+      // postMessage back to the origin frame that the request was cancelled.
+      this._requestInfo.source.postMessage({
+        rpcId: this._requestInfo.rpcId,
+        error: this._error.get() || "User cancelled request",
+      }, this._requestInfo.origin);
+    }
+  }
+
+  hexValueForWellKnownTagId(decimalString) {
```

Comment:
**tips hat**

Integrated with gentle modification (older browsers may lack `String.prototype.startsWith`) in 9e433d2a4f6d1878810ea58e4ed5c0c5ec98b93a